mithril-client/src/certificate_client.rs

```diff
@@ -0,0 +1,513 @@
+    pub async fn list(&self) -> MithrilResult<Vec<MithrilCertificateListItem>> {
+        let response = self
+            .aggregator_client
+            .get_content(AggregatorRequest::ListCertificates)
+            .await
+            .with_context(|| "CertificateClient can not get the certificate list")?;
+        let items = serde_json::from_str::<Vec<MithrilCertificateListItem>>(&response)
+            .with_context(|| "CertificateClient can not deserialize certificate list")?;
+
+        Ok(items)
+    }
+
+    /// Get a single certificate full information from the aggregator.
+    pub async fn get(&self, certificate_hash: &str) -> MithrilResult<Option<MithrilCertificate>> {
+        self.retriever.get(certificate_hash).await
+    }
+
+    /// Validate the chain starting with the certificate with given `certificate_hash`, return the certificate if
+    /// chain valid.
+    ///
+    /// This method will fails if no certicate exists with the given `certificate_hash`.
+    pub async fn verify_chain(&self, certificate_hash: &str) -> MithrilResult<MithrilCertificate> {
+        let certificate = self.retriever.get(certificate_hash).await?.ok_or(anyhow!(
+            "No certificate exist for hash '{certificate_hash}'"
+        ))?;
+
+        self.verifier
+            .verify_chain(&certificate)
+            .await
+            .with_context(|| {
+                format!("Certicate chain of certificate '{certificate_hash}' is invalid")
+            })?;
+
+        Ok(certificate)
+    }
+}
+
+/// Internal type to implement the [InternalCertificateRetriever] trait and avoid a circular
+/// dependency between the [CertificateClient] and the [MithrilCertificateVerifier] that need
+/// a [CertificateRetriever] as a dependency.
+struct InternalCertificateRetriever {
+    aggregator_client: Arc<dyn AggregatorClient>,
+    logger: Logger,
+}
+
+impl InternalCertificateRetriever {
+    async fn get(&self, certificate_hash: &str) -> MithrilResult<Option<MithrilCertificate>> {
+        let response = self
+            .aggregator_client
+            .get_content(AggregatorRequest::GetCertificate {
+                hash: certificate_hash.to_string(),
+            })
+            .await;
+
+        match response {
+            Err(AggregatorClientError::RemoteServerLogical(_)) => Ok(None),
+            Err(e) => Err(e.into()),
+            Ok(response) => {
+                let message =
+                    serde_json::from_str::<CertificateMessage>(&response).map_err(|e| {
+                        crit!(
+                            self.logger,
+                            "Could not create certificate from API message: {e}."
+                        );
+                        debug!(self.logger, "Certificate message = {response}");
+                        e
+                    })?;
+
+                Ok(Some(message))
+            }
+        }
+    }
+}
+
+/// Implementation of a [CertificateVerifier] that can send feedbacks using
```

Comment:
I don't think that's it's possible to remove this struct since the common verifier works on `Certificate` but here we needs to works on `CertificateMessage`.
What can be possible, and I think that we both agree on that, is to simplify the merge most of this verifier `verify_chain` code into the common verifier using a callback as you suggest.
So I can add a TODO but it would be on the `verify_chain` method below.